lib/sgd.hpp

```diff
@@ -0,0 +1,71 @@
+// Copyright 2016 Husky Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cmath>
+#include <functional>
+#include <utility>
+#include <vector>
+
+#include "core/engine.hpp"
+#include "lib/feature_label.hpp"
+#include "lib/gradient_descent.hpp"
+#include "lib/parameter.hpp"
+#include "lib/vector_linalg.hpp"
+
+namespace husky {
+namespace ML {
+
+using lib::AggregatorFactory;
+using vec_double = std::vector<double>;
+using vec_sp = std::vector<std::pair<int, double>>;
+
+// base class for stochastic gradient descent, extends GradientDescentBase
+template <typename ObjT = FeatureLabel, typename ParamT = ParameterBucket<double>>
+class SGD : public GradientDescentBase<ObjT> {
+   private:
+    using ObjL = ObjList<ObjT>;
+
+   public:
+    // constructors
+    SGD() : GradientDescentBase<ObjT>() {}
+    explicit SGD(std::function<vec_sp(ObjT&, std::function<double(int)>)> _gradient_func, double _learning_rate)
+        : GradientDescentBase<ObjT>(_gradient_func, _learning_rate) {}
+
+    void update_vec(ObjL& data, ParamT param_list, int num_global_samples) override {
+        ASSERT_MSG(this->learning_rate_ != 0, "Learning rate is set to 0.");
+        ASSERT_MSG(this->gradient_func_ != nullptr, "Gradient function is not specified.");
+
+        int num_local_samples = data.get_size();              // number of local samples
+        vec_double current_vec = param_list.get_all_param();  // local copy of parameter
+        auto& ac = AggregatorFactory::get_channel();
+
+        list_execute(data, {}, {&ac}, [&grad_func = this->gradient_func_, &rate = this->learning_rate_,
```

Comment:
The only difference is that for FGD current_vec is not changed in list_execute but it is changed for SGD... What is your suggestion?